src/Data/Singletons/Single.hs

```diff
@@ -537,6 +557,82 @@ singPat _var_proms DWildPa = return DWildPa
 -- calls. Specifically, DON'T do the applySing stuff. Just use sError, which
 -- has a custom type (Sing x -> a) anyway.
 
+-- Note [Singling pattern signatures]
+-- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+-- We want to single a pattern signature, like so:
+--
+--   f :: Maybe a -> a
+--   f (Just x :: Maybe a) = x
+--
+-- Naïvely, one might expect this to single straightfowardly as:
+--
+--   sF :: forall (z :: Maybe a). Sing z -> Sing (F z)
+--   sF (SJust sX :: Sing (Just x :: Maybe a)) = sX
+--
+-- But the way GHC typechecks patterns prevents this from working, as GHC won't
+-- know that the type `z` is actually `Just x` until /after/ the entirety of
+-- the `SJust sX` pattern has been typechecked. (See Trac #12018 for an
+-- extended discussion on this topic.)
+--
+-- To work around this design, we resort to a somewhat unsightly trick:
+-- immediately after matching on all the patterns, we perform a case on every
+-- pattern with a pattern signature, like so:
+--
+--   sF :: forall (z :: Maybe a). Sing z -> Sing (F z)
+--   sF (SJust sX :: Sing z)
```

Comment:
I'm worried about this approach. Whenever I do this trick, I do something like

```
f x@(T ...) = case x of (T ... :: ...) ->
```

Note the as-pattern. This means that GHC will always be convinced that the two scrutinees are the same. Here, you're doing it with type signatures, but I'm not sure that's always going to be enough. What if there is kind polymorphism? Ambiguity? Maybe I'm worried about nothing, but this doesn't smell right.

Of course, to make this work, we'd need as-patterns, which th-desugar doesn't support. So perhaps we should generate plain old TH.

Reply:
On the contrary, I _do_ think this is always going to be enough. All that we're doing here is informing the typechecker that the `z` in `(SJust sX :: Sing z)` is equal to `(Just x :: Maybe a)`. The actual choice of expression that we scrutinize isn't important, since we take care to annotate both the original pattern and the scrutinized expression with `Sing z`, so there will never be ambiguity in that department.

Reply:
Hm. That's a convincing response. OK. Proceed.